src/psyclone/psyir/nodes/directive.py

```diff
@@ -133,6 +136,38 @@ def clauses(self):
             return self.children[1:]
         return []
 
+    def gen_post_region_code(self, parent):
+        '''
+        Generates any code that must be executed immediately after the end of
+        the region defined by this directive.
+
+        TODO #1648 this method is only used by the gen_code() code-generation
+        path and should be replaced by functionality in a
+        'lower_to_language_level' method in an LFRic-specific subclass
+        of the appropriate directive.
+
+        :param parent: where to add new f2pygen nodes.
+        :type parent: :py:class:`psyclone.f2pygen.BaseGen`
+
+        '''
+        if not Config.get().distributed_memory or self.ancestor(Loop):
+            return
+
+        commented = False
+        for loop in self.walk(Loop):
+            if not isinstance(loop.parent, Loop):
+                if not commented and loop.unique_modified_args("gh_field"):
+                    commented = True
+                    parent.add(CommentGen(parent, ""))
+                    parent.add(CommentGen(parent,
+                                          " Set halos dirty/clean for fields "
+                                          "modified in the above loop(s)"))
+                    parent.add(CommentGen(parent, ""))
+                loop.gen_mark_halos_clean_dirty(parent)
+
+        if commented:
+            parent.add(CommentGen(parent, ""))
```

Comment:
Could you also add a comment stating that this is the end of the set dirty/clean section? (that might be useful for debugging if we ever should get halos created from different places?)

Reply:
Oh yes, good idea.